### PR TITLE
DPL-831: Refactor sample sheets to use existing messaging config structure

### LINKED ISSUE #1058: DPL-831: Refactor sample sheets to better handle multiple versions, instruments, and libraries
**Describe the Housekeeping**
Refactor sample sheets to better handle multiple versions, instruments, and libraries.

**Blocking issues**
- #1067
- #1085 

**Additional context**
As a follow-on to #993, the proposed solution (#1050) leaves room for improvement to maintain the quality of code going forward. An alternative implementation (#1057) provides additional ideas, but more work is required to form a simpler and more elegant solution.

This has resulted in #1067, part 1 - sample sheets to address the creation of the CSV.  
Part 2 of the solution addresses the instrument and version conflicts introduced by SMRT Link v12 for Sequel IIe.

### PR DESCRIPTION
**Pull request by StephenHulme**
Closes #1058 (part 1)

Changes proposed in this pull request:

- Refactor sample sheets to use existing messaging config structure
- Use a multi-stage abstraction to compile down the Data -> JSON -> CSV
- Refactor the config parsing into a common class for both messaging and sample sheets
- CSVs return a nil field instead of an empty string for nil data
- Update documentation

Dependencies:
- #1091

Sample Sheet Changes:

[sample-sheets-diff.pdf](https://github.com/sanger/traction-service/files/12464051/sample-sheets-diff.pdf)

created with:  
_./scripts/download-sample-sheets $TRACTION_DEV_HOST 1 8 new-style_  
_git diff --no-index --word-diff=color --word-diff-regex=. storage/deprecated storage/new-style --color | [ansi2html](http://www.pixelbeat.org/scripts/ansi2html.sh) | [wkhtmltopdf](https://wkhtmltopdf.org/) - sample-sheets-diff.pdf_

### CODE REVIEW

File: app/csv_generator/pacbio_sample_sheet_message.rb
Comment (stevieing) [inline app/csv_generator/pacbio_sample_sheet_message.rb:15]: This is recursive so maybe you can turn it into a recursive function? As long as you have a terminator it should be fine? It would also make Robocop happy.

File: config/pipelines/pacbio.yml
Comment (stevieing) [inline config/pipelines/pacbio.yml:563]: I suppose we could also add order as an attribute for each field but not sure which is best?

File: app/models/concerns/sample_sheet.rb
Comment (stevieing) [inline app/models/concerns/sample_sheet.rb:27]: might be worth waiting until we merge develop but maybe we could have a method with an argument in the configuration. That would make it much more dynamic. Then for plate 1 and 2 we could have the same method but with a different argument?

File: app/exchanges/README.md
Comment (BenTopping) [inline app/exchanges/README.md:9]: I'd make this less specific as we don't support just v12. Maybe something like 'The required format of the sample sheets can be found in the SMRT Link User Guide' (https://www.pacb.com/support/documentation/)

File: app/exchanges/README.md
Comment (BenTopping) [inline app/exchanges/README.md:13]: I tend to not include internal addresses in the repo's if I can avoid it just as a precautionary measure.

File: app/exchanges/message/message.rb
Comment (BenTopping) [inline app/exchanges/message/message.rb:9]: No sure this example provides much help here.

File: app/exchanges/run_csv/deprecated_pacbio_sample_sheet.rb
Comment (BenTopping) [inline app/exchanges/run_csv/deprecated_pacbio_sample_sheet.rb:1]: Is this needed? If the issue is the structure of the config for v11 sample sheet, I would change the sample sheet config to work with the new method rather than support the old method if possible.

Comment (BenTopping) [reply app/exchanges/run_csv/deprecated_pacbio_sample_sheet.rb:1]: Agreed v10 and v11 support for new generation methods can be moved to a separate story / pr to avoid scope creep

**Review by BenTopping — CHANGES_REQUESTED** [review]
I haven't pulled down locally to check yet. Just a few comments around documentation mainly.

**Review by BenTopping — APPROVED** [review]

**Pull request closed without merging**
